File: Cpp_Parallel/src/lib/ParallelSelectionSort.cpp
#include "include/ParallelSelectionSort.h";

void ParallelSelectionSort::parallelSelectionSort(std::vector<int>& array, int numThreads) {
    std::vector<std::thread> threads;
    int n = array.size();

    for (int i = 0; i < n - 1; i++) {
        int outerIndex = i;
        threads.emplace_back([=, &array]() {
            selectionSortSegment(array, outerIndex + 1, n);
            std::swap(array[outerIndex], array[outerIndex + 1]);
        });
    }

    for (auto& thread : threads) {
        thread.join();
    }
}

void ParallelSelectionSort::selectionSortSegment(std::vector<int>& array, int start, int end) {
    int currentMinIndex = start;
    for (int j = start; j < end; j++) {
        if (array[j] < array[currentMinIndex]) {
            currentMinIndex = j;
        }
    }
    std::swap(array[start], array[currentMinIndex]);
}

File: Java_Parallel/src/ParallelSorts/ParallelSelectionSort.java
package ParallelSorts;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class ParallelSelectionSort {

    public static void parallelSelectionSort(int[] array, int numThreads) {
        ExecutorService executor = Executors.newFixedThreadPool(numThreads);
        int n = array.length;

        for (int i = 0; i < n - 1; i++) {
            final int outerIndex = i;
            executor.submit(() -> {
                int currentMinIndex = outerIndex;
                for (int j = outerIndex + 1; j < n; j++) {
                    if (array[j] < array[currentMinIndex]) {
                        currentMinIndex = j;
                    }
                }
                swap(array, outerIndex, currentMinIndex);
            });
        }
        executor.shutdown();
        while (!executor.isTerminated()) {} // Wait for all tasks to finish
    }

    private static void swap(int[] array, int i, int j) {
        int temp = array[i];
        array[i] = array[j];
        array[j] = temp;
    }
}
